node/pkg/tss/fault_tolerance.go

```diff
@@ -0,0 +1,397 @@
+	// sigs that once the revive time expire should be retried.
+	retrySigs                   map[party.Digest]*signatureState
+	liveSigsWaitingForThisParty map[party.Digest]*signatureState
+}
+
+// Describes a specfic party's data in terms of fault tolerance.
+type ftParty struct {
+	partyID     *tss.PartyID
+	ftPartyData map[vaa.ChainID]*ftChainData
+}
+
+type ftTracker struct {
+	sigAlerts internal.Ttlheap[*signatureState]
+	sigsState map[party.Digest]*signatureState
+	// for starters, we assume any fault is on all chains.
+	membersData map[strPartyId]*ftParty
+}
+
+func newEmptyChainData() *ftChainData {
+	return &ftChainData{
+		timeToRevive:                time.Now(),
+		retrySigs:                   map[party.Digest]*signatureState{},
+		liveSigsWaitingForThisParty: map[party.Digest]*signatureState{},
+	}
+}
+
+// a single threaded env, that inspects incoming signatures request, message deliveries etc.
+func (t *Engine) ftTracker() {
+	f := &ftTracker{
+		sigAlerts:   internal.NewTtlHeap[*signatureState](),
+		sigsState:   make(map[party.Digest]*signatureState),
+		membersData: make(map[strPartyId]*ftParty),
+	}
+
+	for _, pid := range t.GuardianStorage.Guardians {
+		strPid := strPartyId(partyIdToString(pid))
+		f.membersData[strPid] = &ftParty{
+			partyID:     pid,
+			ftPartyData: map[vaa.ChainID]*ftChainData{},
+		}
+	}
+
+	for {
+		select {
+		case <-t.ctx.Done():
+			return
+
+		case cmd := <-t.ftCommandChan:
+			f.executeCommand(t, cmd)
+		case <-f.sigAlerts.WaitOnTimer():
+			f.inspectAlertHeapsTop(t)
+		}
+	}
+}
+
+func (f *ftTracker) executeCommand(t *Engine, cmd ftCommand) {
+	switch c := cmd.(type) {
+	case *signCommand:
+		f.executeSignCommand(t, c)
+	case *deliveryCommand:
+		f.executeDeliveryCommand(t, c)
+	case *getInactiveGuardiansCommand:
+		f.executeGetIncativeGuardiansCommand(t, c)
+	case *parsedProblem:
+		f.executeParsedProblemCommand(t, c)
+	default:
+		t.logger.Error("received unknown command type", zap.Any("cmd", cmd))
+	}
+}
+
+func (f *ftTracker) executeParsedProblemCommand(t *Engine, cmd *parsedProblem) {
+	t.logger.Info("received a problem message from another guardian", zap.Any("problem issuer", cmd.issuer))
+
+	pid := protoToPartyId(cmd.issuer)
+
+	m := f.membersData[strPartyId(partyIdToString(pid))]
+
+	reviveTime := time.Now().Add(t.GuardianStorage.GuardianSigningDownTime)
+	chainID := vaa.ChainID(cmd.ChainID)
+
+	chainData, ok := m.ftPartyData[chainID]
+	if !ok {
+		chainData = newEmptyChainData()
+		m.ftPartyData[chainID] = chainData
+	}
+
+	// we update the revival time only if the revival time had passed
+	if time.Now().After(chainData.timeToRevive) {
+		chainData.timeToRevive = reviveTime
+		// TODO: insert to some timed heap
+	}
+
+	if equalPartyIds(pid, t.Self) {
+		return
+	}
+
+	retryNow := chainData.liveSigsWaitingForThisParty
+	chainData.liveSigsWaitingForThisParty = map[party.Digest]*signatureState{} // clear the live sigs.
+
+	go func() {
```

Comment:
You are mostly correct. I'll some comments. 
I let the party learn and update the state about itself so it'll have correct values when it attempts to sign in the future. 
I return before I set up the retry since this server wouldn't be signing in any case. 

I restart the protocol for live signatures where this guardian is part of the committee only. 
I need to ensure they are cleaned once a sig is produced + some time had passed.

Right now I'm not removing guardians with missing heartbeats. I still need to implement it (probably collect valid heartbeats, and add timer that runs in fixed interval to check no heartbeat is too old ).

regarding your suggestion for a channel, I'm not sure I follow, 
I didn't want to run it inside the ftTracker to ensure it wont be blocked somehome by heavy processing. 
But I can create a goroutine and a channel to reuse, and let it call the retry calls.